src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,136 @@
+from pathlib import Path
 from google.cloud import bigquery
 import argparse
-import random
+import numpy as np
 import anndata as ad
+from scipy.sparse import coo_matrix
 
+class Cell:
+    def __init__(self, cas_cell_index, original_cell_id, cell_type):
+        self.cas_cell_index = cas_cell_index
+        self.original_cell_id = original_cell_id
+        self.cell_type = cell_type
 
-# assumes that cas_cell_info.cas_feature_index values are a contiguous list of ints starting at 0
-def get_random_ids(project, dataset, client, num_cells):
-    query = client.query(f"SELECT MAX(cas_cell_index) AS max_table_number FROM `{project}.{dataset}.cas_cell_info`")
-    max_cell_id = int([row.max_table_number for row in list(query.result())][0])
-    print(f"Getting {num_cells} random IDs between 0 and {max_cell_id}...")
-    cell_ids = list(range(0, max_cell_id + 1))
-    random.shuffle(cell_ids)
-    del cell_ids[num_cells:]
-    return cell_ids
+class Feature:
+    def __init__(self, cas_feature_index, original_feature_id, feature_name):
+        self.cas_feature_index = cas_feature_index
+        self.original_feature_id = original_feature_id
+        self.feature_name = feature_name
 
+# Returns an array of <num_cells> cas_cell_index values pulled at random from the cas_cell_info table
+def get_random_cell_ids(project, dataset, client, num_cells):
+    query = client.query(f"SELECT cas_cell_index, rand() AS rand_val FROM `{project}.{dataset}.cas_cell_info` ORDER BY rand_val LIMIT {num_cells}")
```

Comment:
nice!

Reply:
I can't take credit for that!  ( @kcibul )